### DISCUSSION TITLE
[mini.hues] My favorite hues and persisting random ones

### DISCUSSION THREAD

**Discussion by pkazmier** [nvim-mini/mini.nvim#919] [Show and tell]
Here are two of my favorite hues thus far.

### Deep Purple

```lua
local hues = require("mini.hues")
vim.g.colors_name = "minihues-purple"
local opts = {
  accent = "azure",
  background = "#151025",
  foreground = "#c0c8cb",
}
hues.setup(opts)
require("plugins.mini.hues").apply_custom_highlights(opts)
```
<img width="2560" alt="Screenshot 2024-05-28 at 8 07 55 AM" src="https://github.com/echasnovski/mini.nvim/assets/747855/d081eb4b-8fef-48c0-887b-fa76a64b1209">

### Olive

```lua
local hues = require("mini.hues")
vim.g.colors_name = "minihues-olive"
local opts = {
  accent = "green",
  background = "#18271a",
  foreground = "#c2c9c3",
}
hues.setup(opts)
require("plugins.mini.hues").apply_custom_highlights(opts)
```
<img width="2560" alt="Screenshot 2024-05-28 at 8 07 02 AM" src="https://github.com/echasnovski/mini.nvim/assets/747855/7617b67a-c2e6-478f-8a72-3b814102f3f2">

## Persisting random themes

I sometimes generate a random hue that I really like—such as the olive one from above:

```lua
local hues = require("mini.hues")
math.randomseed(vim.loop.hrtime())
local base_colors = hues.gen_random_base_colors()
local opts = {
  background = base_colors.background,
  foreground = base_colors.foreground,
  saturation = vim.o.background == "dark" and "medium" or "high",
  accent = "azure",
}
hues.setup(opts)
require("plugins.mini.hues").apply_custom_highlights(opts)
vim.g.colors_name = "minihues-myrandomhue"
```

When I find one that I like, I want to be able to persist it for future use, so I wrote an exporter that will generate a new colorscheme with my custom highlights as well as a matching wezterm colorscheme (because you are a monster if your terminal and neovim themes aren't in sync 😄 ).

<img width="792" alt="Screenshot 2024-05-28 at 8 14 58 AM" src="https://github.com/echasnovski/mini.nvim/assets/747855/a45b2923-50f5-4a10-b177-2bbf08fe868a">
<img width="792" alt="Screenshot 2024-05-28 at 8 15 08 AM" src="https://github.com/echasnovski/mini.nvim/assets/747855/a3bb2539-3978-4ae8-a322-c8ef6033b0cf">

My exporter opens two buffers with the rendered themes in case I want to edit them before saving. Typically I want to change the accent color based on the hue. The wezterm config is specific to a [dynamic configuration selector](https://github.com/wez/wezterm/discussions/5435) I recently wrote for my own wezterm setup that allows me to dynamically fuzzy select fonts, colorschemes, etc ...

```lua
local M = {}
local H = {}

M.export_minihues_theme = function()
  local ok, theme_name = pcall(vim.fn.input, {
    prompt = "Enter name for color scheme: minihues-",
    cancelreturn = false,
  })
  if not ok or theme_name == false then
    return nil
  end
  local filename = "minihues-" .. theme_name
  local p = require("mini.hues").make_palette({})
  H.render(H.minihues_path(filename), H.minihues_template, filename, vim.inspect(H.minihues_opts(p)))
  vim.cmd("split")
  local display_name = "MiniHues " .. theme_name:gsub("^%l", string.upper)
  H.render(H.wezterm_path(filename), H.wezterm_template, display_name, vim.inspect(H.wezterm_opts(p)))
end

H.wezterm_path = function(theme_name)
  local dir = os.getenv("WEZTERM_CONFIG_DIR")
  return string.format("%s/colorschemes/%s.lua", dir, theme_name)
end

H.minihues_path = function(theme_name)
  local dir = string.format("~/.config/%s/colors", os.getenv("NVIM_APPNAME") or "nvim")
  return string.format("%s/%s.lua", dir, theme_name)
end

H.render = function(filename, template, ...)
  local rendered = string.format(template, ...)
  vim.cmd(string.format("edit %s", filename))
  vim.api.nvim_buf_set_lines(0, 0, -1, false, vim.split(rendered, "\n"))
end

H.minihues_template = [[local hues = require("mini.hues")
vim.g.colors_name = "%s"
local opts = %s
hues.setup(opts)
require("plugins.mini.hues").apply_custom_highlights(opts)
]]

H.wezterm_template = [[local M = {}
local name = "%s"

M.init = function()
  return name
end

M.activate = function(config)
  config.color_schemes = config.color_schemes or {}
  config.color_schemes[name] = %s
  config.color_scheme = name
end

return M
]]

H.minihues_opts = function(p)
  return {
    background = p.bg,
    foreground = p.fg,
    accent = "azure",
  }
end

H.wezterm_opts = function(p)
  return {
    foreground = p.fg,
    background = p.bg,
    cursor_bg = p.azure,
    cursor_fg = p.azure_bg,
    cursor_border = p.azure,
    selection_bg = p.bg_mid2,
    scrollbar_thumb = p.bg_mid,
    split = p.bg_mid2,
    ansi = {
      p.bg_mid,
      p.red,
      p.green,
      p.yellow,
      p.azure,
      p.purple,
      p.cyan,
      p.fg_mid2,
    },
    brights = {
      p.bg_mid2,
      p.red,
      p.green,
      p.yellow,
      p.azure,
      p.purple,
      p.cyan,
      p.fg_mid,
    },
    indexed = { [136] = p.orange },
    compose_cursor = p.orange,
    copy_mode_active_highlight_bg = { Color = p.bg },
    copy_mode_active_highlight_fg = { Color = p.fg_mid },
    copy_mode_inactive_highlight_bg = { Color = p.green },
    copy_mode_inactive_highlight_fg = { Color = p.bg },
    quick_select_label_fg = { Color = p.bg },
    quick_select_label_bg = { Color = p.orange },
    quick_select_match_bg = { Color = p.bg_mid },
    quick_select_match_fg = { Color = p.azure },
  }
end

return M
```

**Comment by echasnovski** [comment]
Both of these solutions look great!

Strangely, I am personally not a fan of purple as a color, but looking at that first screenshot made me think "Those are beautiful colors!" :)

**Reply by pkazmier** [reply]
😄 It’s amazing how a nice wallpaper can augment the color scheme. I created those when I saw that recent post on Reddit for that neofusion theme. I realized I needed to create wallpapers for my mini.hues favorites, so  I spent hours on my day off yesterday playing around with different colors in my Affinity Photo. I love the way the purple looked, so I did some advertising for mini in [/r/vimporn](https://www.reddit.com/r/vimporn/comments/1d211kn/mininvim_configuration/) last night.

**Reply by echasnovski** [reply]
Yeah, simple gradient with proper colors goes a long way.
